<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><style>body {
  max-width: 980px;
  border: 1px solid #ddd;
  outline: 1300px solid #fff;
  margin: 16px auto;
}

body .markdown-body
{
  padding: 45px;
}

@font-face {
  font-family: fontawesome-mini;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAABE0AA8AAAAAHWwAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABHU1VCAAABWAAAADsAAABUIIslek9TLzIAAAGUAAAAQwAAAFY3d1HZY21hcAAAAdgAAACqAAACOvWLi0FjdnQgAAAChAAAABMAAAAgBtX/BGZwZ20AAAKYAAAFkAAAC3CKkZBZZ2FzcAAACCgAAAAIAAAACAAAABBnbHlmAAAIMAAABdQAAAjkYT9TNWhlYWQAAA4EAAAAMwAAADYQ6WvNaGhlYQAADjgAAAAfAAAAJAc6A1pobXR4AAAOWAAAACAAAAA0Kmz/7mxvY2EAAA54AAAAHAAAABwQPBJubWF4cAAADpQAAAAgAAAAIAEHC/NuYW1lAAAOtAAAAYQAAALxhQT4h3Bvc3QAABA4AAAAfgAAAMS3SYh9cHJlcAAAELgAAAB6AAAAhuVBK7x4nGNgZGBg4GIwYLBjYHJx8wlh4MtJLMljkGJgYYAAkDwymzEnMz2RgQPGA8qxgGkOIGaDiAIAJjsFSAB4nGNgZHZmnMDAysDAVMW0h4GBoQdCMz5gMGRkAooysDIzYAUBaa4pDA4Pwz+yMwf9z2KIYg5imAYUZgTJAQDcoQvQAHic7ZHNDYJAFIRnBXf94cDRIiyCKkCpwFCPJ092RcKNDoYKcN4+EmMPvpdvk539zQyAPYBCXEUJhBcCrJ5SQ9YLnLJe4qF5rdb+uWPDngNHTkta101pNyWa8lMhn6xx2dqUnW4q9YOIhAOOeueMSgsR/6ry+P7O5s6xVNg4chBsHUuFnWNJ8uZYwrw7chrsHXkODo7cB0dHOYCTY8kv0VE2WJKD6gOlWjsxAAB4nGNgQAMSEMgc9D8LhAESbAPdAHicrVZpd9NGFB15SZyELCULLWphxMRpsEYmbMGACUGyYyBdnK2VoIsUO+m+8Ynf4F/zZNpz6Dd+Wu8bLySQtOdwmpOjd+fN1czbZRJaktgL65GUmy/F1NYmjew8CemGTctRfCg7eyFlisnfBVEQrZbatx2HREQiULWusEQQ+x5ZmmR86FFGy7akV03KLT3pLlvjQb1V334aOsqxO6GkZjN0aD2yJVUYVaJIpj1S0qZlqPorSSu8v8LMV81QwohOImm8GcbQSN4bZ7TKaDW24yiKbLLcKFIkmuFBFHmU1RLn5IoJDMoHzZDyyqcR5cP8iKzYo5xWsEu20/y+L3mndzk/sV9vUbbkQB/Ijuzg7HQlX4RbW2HctJPtKFQRdtd3QmzZ7FT/Zo/ymkYDtysyvdCMYKl8hRArP6HM/iFZLZxP+ZJHo1qykRNB62VO7Es+gdbjiClxzRhZ0N3RCRHU/ZIzDPaYPh788d4plgsTAngcy3pHJZwIEylhczRJ2jByYCVliyqp9a6YOOV1WsRbwn7t2tGXzmjjUHdiPFsPHVs5UcnxaFKnmUyd2knNoykNopR0JnjMrwMoP6JJXm1jNYmVR9M4ZsaERCICLdxLU0EsO7GkKQTNoxm9uRumuXYtWqTJA/Xco/f05la4udNT2g70s0Z/VqdiOtgL0+lp5C/xadrlIkXp+ukZfkziQdYCMpEtNsOUgwdv/Q7Sy9eWHIXXBtju7fMrqH3WRPCkAfsb0B5P1SkJTIWYVYhWQGKta1mWydWsFqnI1HdDmla+rNMEinIcF8e+jHH9XzMzlpgSvt+J07MjLj1z7UsI0xx8m3U9mtepxXIBcWZ5TqdZlu/rNMfyA53mWZ7X6QhLW6ejLD/UaYHlRzodY3lBC5p038GQizDkAg6QMISlA0NYXoIhLBUMYbkIQ1gWYQjLJRjC8mMYwnIZhrC8rGXV1FNJ49qZWAZsQmBijh65zEXlaiq5VEK7aFRqQ54SbpVUFM+qf2WgXjzyhjmwFkiXyJpfMc6Vj0bl+NYVLW8aO1fAsepvH472OfFS1ouFPwX/1dZUJb1izcOTq/Abhp5sJ6o2qXh0TZfPVT26/l9UVFgL9BtIhVgoyrJscGcihI86nYZqoJVDzGzMPLTrdcuan8P9NzFCFlD9+DcUGgvcg05ZSVnt4KzV19uy3DuDcjgTLEkxN/P6VvgiI7PSfpFZyp6PfB5wBYxKZdhqA60VvNknMQ+Z3iTPBHFbUTZI2tjOBIkNHPOAefOdBCZh6qoN5E7hhg34BWFuwXknXKJ6oyyH7kXs8yik/Fun4kT2qGiMwLPZG2Gv70LKb3EMJDT5pX4MVBWhqRg1FdA0Um6oBl/G2bptQsYO9CMqdsOyrOLDxxb3lZJtGYR8pIjVo6Of1l6iTqrcfmYUl++dvgXBIDUxf3vfdHGQyrtayTJHbQNTtxqVU9eaQ+NVh+rmUfW94+wTOWuabronHnpf06rbwcVcLLD2bQ7SUiYX1PVhhQ2iy8WlUOplNEnvuAcYFhjQ71CKjf+r+th8nitVhdFxJN9O1LfR52AM/A/Yf0f1A9D3Y+hyDS7P95oTn2704WyZrqIX66foNzBrrblZugbc0HQD4iFHrY64yg18pwZxeqS5HOkh4GPdFeIBwCaAxeAT3bWM5lMAo/mMOT7A58xh0GQOgy3mMNhmzhrADnMY7DKHwR5zGHzBnHWAL5nDIGQOg4g5DJ4wJwB4yhwGXzGHwdfMYfANc+4DfMscBjFzGCTMYbCv6dYwzC1e0F2gtkFVoANTT1jcw+JQU2XI/o4Xhv29Qcz+wSCm/qjp9pD6Ey8M9WeDmPqLQUz9VdOdIfU3Xhjq7wYx9Q+DmPpMvxjLZQa/jHyXCgeUXWw+5++J9w/bxUC5AAEAAf//AA94nIVVX2hbZRQ/5/t7893s5ja9f7ouzdZ0TTqz3bRJmogbWya6bG6Cq0VbSV2ddIJjFtfIQHEig80Hda8yUN/0YQz8AyriiyD+xQd92R4HCnaCb3samnpumrpsCsLlfPf7zvedc37nL3CAtc/5W/wQZGA3tOBSY/g+TMjHmwzEoM1Q8+ZjRZY4oJhmBw5/YB6Za0yC5AkhlwA1A1yCBIBOwCII0Cj0U8BAMdUCzq05sKwkP7SlUY6fcJk4Fb/RyE79/6P5hjM/F4aZiXBoeMgzcqQ4Xi1hPqfDLG5FT+lchCVU3lYMyvuwhl1mqndQL0RsuloLywHtthLXI06OblTrhfWVnpSJ5+mwu/JdbtuN3IAnkW0LLMcRwaC7ktrlzridM6kVdyf9uO1UNBByI7JhwtG2sEwab07ORBeilWhqavJCqV0qzZTOl/7ZXQ5TbTcdcFelyGhhRDAQpdqp1FEX3w3cFTc1k9pJQkmm4ySCbSikxRP2QOfN+0tHS5MrpQuTU1Mk5nw0E5Xa0WvrOwDyGax9yB9ma6DAg82wHc43SAGTI4GjBWebOePAERFE8/AHaQpZASSTy8A4WwZiLQMQ82mFKATO0ILicRAoDm9p5P99E5b/fXG+kQYY3TYUuqmERWYoT0u/GNYL2q/4WB3LaVS+VynXsVYIcWw6DkCh3nX1D+VzlYN4LClF5yexSQos8exqZ3KVP+wtrC54u4Nznq6cq+xpMpUUnZ8FUYzE86ud0g28NOIv3Gj5/rmA3ABs7S/ywzFuQ4qyd6QxfNtiQIaEgp3w/entQg4Vcbqa16M5FfpeUB8t1+qeg7mI7cUyOe79wOk86gSxkVec4KPTX69++5x68Yubn5/F+w52z7u08sJX7fZXv8ekT/d2mILJxq6sn+SC6qEJknzLJCxyZEKwWVqYmAPBxBE/9DLeZiWHu7lcr/VytrCRuHojncNuTt9h46tmacmYisnSamdN2bZptcsmSysdVsy1PrOvOzF3xN64Rb937t/og9KHxYdcjIUqFAmIAHGHNzlns+RTPgeUYAQm9DwpNxfxbhhBHPaw3/gfTcXO2L+eJVIx5nsyGkvm9X4/f+bGkH45G0PaSjcMXTjcZyTvi3UdHoCDjQd3IDUVsgwYmUoJK/gp4JJxeRI0MKHZIkgynyIBqBTOUs6rOVCojvjZ4mCQz49ZMlMcp8QoYk6NoBfsxnJtsBohpa8iGJS+ZH7gU7NxME6cmF+t7cO9vB8d3jTWSct0ycW9ranXmolNDwmVkNnxe+8JtoztwS5rKJ0xWS95tQ/1zMYzg69MzUZnNtl1ofNbsml/OJm6f9wjRjpnu2o4MzHzn77IQkRd+1DjwMQ2pqSjGMMhyjrgTbBAKksuUm0iU7hI0aN2wOKOq7WYBSH0HGihj/jkiPxAfmwsEbfYrjMG+j3ij932Db/LV7I/xruNrhnroxjR9HRMb2nTvO0ZXOoHPk8H2ZhDPx93qcE/53sH5np/dkIP7zzhTVKdR/BAY/9ElkkR+A6lJGsqpJ4oQcTxpvBT3Kn58VkaJjgHyPEIws57xkaHh9KuVpDEpJZeMbZ5w/zBHi5NMQ4r5VphsFqID7TyB9eR4pX216c3AHxpdAwoqU9qg0ZJ6yVLKmMSz1iG2z27ifx18NkY0LPx1W/wCc2l5LrznrIsiKsqbmB78A9wIGx4tI8rjihVHJyY9pgMirenVq0yWg7Iw7eogG7ZgYM3qR9959A/fZkg6MnD/exlkmc+jWV4SB15XUR+eqC6l6ZmgPtN9z5JMfik05OV8ljylunJ4J+wA/FUaQSSKotsYsCWqaPBidBLcxkWx7XKFRIb45TGaEhjlF9uUVPqXOtcIwsXbBvfoZXIyRYFdkfnqjExH98xpnPczqzjX/uNdO1Y17Wpi5+6Ts8BXtjVFasp9KZ1mOiNbH65c5w6HgmyF2jFCZywM8mWjRc7T5Pmt0lRy7Y71+jYbpGyvwG4sH0XeJxjYGRgYADiwBB/53h+m68M3MwvgCIM1z5N/g6j///9v5H5BbMnkMvBwAQSBQCIcA9gAHicY2BkYGAO+p8FJF/8//v/F/MLBqAICuAFALYQB5kAeJxjfsHAwLwAiCNB+P9fbJjJmoGBMRUo/wKCAfO2EnQAAAAAANoBXgGcAgICVALaA1IDvAPkBAYEPARyAAEAAAANAF0ABAAAAAAAAgAUACQAcwAAAG4LcAAAAAB4nHWRzWrCQBSFT+pPqUIXLXTTzayKUohGKIibCoLuhbrrYtTRxCYZmYyKyz5Fd32HvlDfoO/QkziIFJtw9bvnnpl7ZwLgBt/wcHieGAf2UGd24Atcou+4RH3kuEweO66QXx1XyaHjGh6ROa7jFp/cwStfMVvhy7GHO+/e8QWuvcBxifqz4zL5xXGF/Oa4Sn53XMPE+3Bcx4P3M9DrvYmWoRWNQVN02kFXTPdCU4pSGQu5saE2meiLhU6timPtz3SSs9ypTCdqrJabWJoT5QQnymSRTkXgt0/UkUqVkVbN807ZdtmxdiEWRidi6HqItdErNbN+aO2612qd9sYAGmvsYRBhyUu0EGhQbfK/gzYCdElTOgSdB1eEFBIxFYkNV4RFJWPeZyyYpVQVHTHZx4y/yVGX2LGWFZri51TccUOn5B7nPefVCSPvGhVVwUl9znveO2KkhV8Wk82PZ8qwZf8OVcu1+fSmWCMw/HMOwXvKaysqM+p+cVuWag8tvv+c+xdd+4+teJxtjUEOwiAURJla24KliQfhUA2g/Sl+CKXx+loNrpzVezOLEY34Ron/0WhwQoszOvQYIKFwwQiNSbSBeO2SZ0tBP4j3zVjKNng32ZmtD1VVXCuOiw/pJ8S3WOU6l+K5UOTaDC4+2TjKMtN9KQf1ezLx/Sg/00FCvABHhjDjAAB4nGPw3sFwIihiIyNjX+QGxp0cDBwMyQUbGVidNjEwMmiBGJu5mBg5ICw+BjCLzWkX0wGgNCeQze60i8EBwmZmcNmowtgRGLHBoSNiI3OKy0Y1EG8XRwMDI4tDR3JIBEhJJBBs5mFi5NHawfi/dQNL70YmBhcADHYj9AAA) format('woff');
}

.markdown-body {
  font-family: sans-serif;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  color: #333333;
  overflow: hidden;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body b,
.markdown-body strong {
  font-weight: bold;
}

.markdown-body mark {
  background: #ff0;
  color: #000;
  font-style: italic;
  font-weight: bold;
}

.markdown-body sub,
.markdown-body sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
.markdown-body sup {
  top: -0.5em;
}
.markdown-body sub {
  bottom: -0.25em;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre,
.markdown-body samp {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body .codehilitetable {
  border: 0;
  border-spacing: 0;
}

.markdown-body .codehilitetable tr {
  border: 0;
}

.markdown-body .codehilitetable pre,
.markdown-body .codehilitetable div.codehilite {
  margin: 0;
}

.markdown-body .linenos,
.markdown-body .code,
.markdown-body .codehilitetable td {
  border: 0;
  padding: 0;
}

.markdown-body td:not(.linenos) .linenodiv {
  padding: 0 !important;
}

.markdown-body .code {
  width: 100%;
}

.markdown-body .linenos div pre,
.markdown-body .linenodiv pre,
.markdown-body .linenodiv {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-left-radius: 3px;
  -webkit-border-bottom-left-radius: 3px;
  -moz-border-radius-topleft: 3px;
  -moz-border-radius-bottomleft: 3px;
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.markdown-body .code div pre,
.markdown-body .code div {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-right-radius: 3px;
  -webkit-border-bottom-right-radius: 3px;
  -moz-border-radius-topright: 3px;
  -moz-border-radius-bottomright: 3px;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.markdown-body * {
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body input {
  font: 13px Helvetica, arial, freesans, clean, sans-serif, "Segoe UI Emoji", "Segoe UI Symbol";
  line-height: 1.4;
}

.markdown-body a {
  color: #4183c4;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:focus,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before,
.markdown-body hr:after {
  display: table;
  content: " ";
}

.markdown-body hr:after {
  clear: both;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code,
.markdown-body pre,
.markdown-body samp {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body kbd {
  background-color: #e7e7e7;
  background-image: -moz-linear-gradient(#fefefe, #e7e7e7);
  background-image: -webkit-linear-gradient(#fefefe, #e7e7e7);
  background-image: linear-gradient(#fefefe, #e7e7e7);
  background-repeat: repeat-x;
  border-radius: 2px;
  border: 1px solid #cfcfcf;
  color: #000;
  padding: 3px 5px;
  line-height: 10px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  display: inline-block;
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body .headerlink {
  font: normal 400 16px fontawesome-mini;
  vertical-align: middle;
  margin-left: -16px;
  float: left;
  display: inline-block;
  text-decoration: none;
  opacity: 0;
  color: #333;
}

.markdown-body .headerlink:focus {
  outline: none;
}

.markdown-body h1 .headerlink {
  margin-top: 0.8rem;
}

.markdown-body h2 .headerlink,
.markdown-body h3 .headerlink {
  margin-top: 0.6rem;
}

.markdown-body h4 .headerlink {
  margin-top: 0.2rem;
}

.markdown-body h5 .headerlink,
.markdown-body h6 .headerlink {
  margin-top: 0;
}

.markdown-body .headerlink:hover,
.markdown-body h1:hover .headerlink,
.markdown-body h2:hover .headerlink,
.markdown-body h3:hover .headerlink,
.markdown-body h4:hover .headerlink,
.markdown-body h5:hover .headerlink,
.markdown-body h6:hover .headerlink {
  opacity: 1;
  text-decoration: none;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre,
.markdown-body .admonition {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body code,
.markdown-body samp {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .codehilite {
  margin-bottom: 16px;
}

.markdown-body .codehilite pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}

.markdown-body .codehilite pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

/* Admonition */
.markdown-body .admonition {
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  position: relative;
  border-radius: 3px;
  border: 1px solid #e0e0e0;
  border-left: 6px solid #333;
  padding: 10px 10px 10px 30px;
}

.markdown-body .admonition table {
  color: #333;
}

.markdown-body .admonition p {
  padding: 0;
}

.markdown-body .admonition-title {
  font-weight: bold;
  margin: 0;
}

.markdown-body .admonition>.admonition-title {
  color: #333;
}

.markdown-body .attention>.admonition-title {
  color: #a6d796;
}

.markdown-body .caution>.admonition-title {
  color: #d7a796;
}

.markdown-body .hint>.admonition-title {
  color: #96c6d7;
}

.markdown-body .danger>.admonition-title {
  color: #c25f77;
}

.markdown-body .question>.admonition-title {
  color: #96a6d7;
}

.markdown-body .note>.admonition-title {
  color: #d7c896;
}

.markdown-body .admonition:before,
.markdown-body .attention:before,
.markdown-body .caution:before,
.markdown-body .hint:before,
.markdown-body .danger:before,
.markdown-body .question:before,
.markdown-body .note:before {
  font: normal normal 16px fontawesome-mini;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 1.5;
  color: #333;
  position: absolute;
  left: 0;
  top: 0;
  padding-top: 10px;
  padding-left: 10px;
}

.markdown-body .admonition:before {
  content: "\f056\00a0";
  color: 333;
}

.markdown-body .attention:before {
  content: "\f058\00a0";
  color: #a6d796;
}

.markdown-body .caution:before {
  content: "\f06a\00a0";
  color: #d7a796;
}

.markdown-body .hint:before {
  content: "\f05a\00a0";
  color: #96c6d7;
}

.markdown-body .danger:before {
  content: "\f057\00a0";
  color: #c25f77;
}

.markdown-body .question:before {
  content: "\f059\00a0";
  color: #96a6d7;
}

.markdown-body .note:before {
  content: "\f040\00a0";
  color: #d7c896;
}

.markdown-body .admonition::after {
  content: normal;
}

.markdown-body .attention {
  border-left: 6px solid #a6d796;
}

.markdown-body .caution {
  border-left: 6px solid #d7a796;
}

.markdown-body .hint {
  border-left: 6px solid #96c6d7;
}

.markdown-body .danger {
  border-left: 6px solid #c25f77;
}

.markdown-body .question {
  border-left: 6px solid #96a6d7;
}

.markdown-body .note {
  border-left: 6px solid #d7c896;
}

.markdown-body .admonition>*:first-child {
  margin-top: 0 !important;
}

.markdown-body .admonition>*:last-child {
  margin-bottom: 0 !important;
}

/* progress bar*/
.markdown-body .progress {
  display: block;
  width: 300px;
  margin: 10px 0;
  height: 24px;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #ededed;
  position: relative;
  box-shadow: inset -1px 1px 3px rgba(0, 0, 0, .1);
}

.markdown-body .progress-label {
  position: absolute;
  text-align: center;
  font-weight: bold;
  width: 100%; margin: 0;
  line-height: 24px;
  color: #333;
  text-shadow: 1px 1px 0 #fefefe, -1px -1px 0 #fefefe, -1px 1px 0 #fefefe, 1px -1px 0 #fefefe, 0 1px 0 #fefefe, 0 -1px 0 #fefefe, 1px 0 0 #fefefe, -1px 0 0 #fefefe, 1px 1px 2px #000;
  -webkit-font-smoothing: antialiased !important;
  white-space: nowrap;
  overflow: hidden;
}

.markdown-body .progress-bar {
  height: 24px;
  float: left;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #96c6d7;
  box-shadow: inset 0 1px 0 rgba(255, 255, 255, .5), inset 0 -1px 0 rgba(0, 0, 0, .1);
  background-size: 30px 30px;
  background-image: -webkit-linear-gradient(
    135deg, rgba(255, 255, 255, .4) 27%,
    transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%,
    transparent 77%, transparent
  );
  background-image: -moz-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -ms-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -o-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
}

.markdown-body .progress-100plus .progress-bar {
  background-color: #a6d796;
}

.markdown-body .progress-80plus .progress-bar {
  background-color: #c6d796;
}

.markdown-body .progress-60plus .progress-bar {
  background-color: #d7c896;
}

.markdown-body .progress-40plus .progress-bar {
  background-color: #d7a796;
}

.markdown-body .progress-20plus .progress-bar {
  background-color: #d796a6;
}

.markdown-body .progress-0plus .progress-bar {
  background-color: #c25f77;
}

.markdown-body .candystripe-animate .progress-bar{
  -webkit-animation: animate-stripes 3s linear infinite;
  -moz-animation: animate-stripes 3s linear infinite;
  animation: animate-stripes 3s linear infinite;
}

@-webkit-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@-moz-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

.markdown-body .gloss .progress-bar {
  box-shadow:
    inset 0 4px 12px rgba(255, 255, 255, .7),
    inset 0 -12px 0 rgba(0, 0, 0, .05);
}

/* MultiMarkdown Critic Blocks */
.markdown-body .critic_mark {
  background: #ff0;
}

.markdown-body .critic_delete {
  color: #c82829;
  text-decoration: line-through;
}

.markdown-body .critic_insert {
  color: #718c00 ;
  text-decoration: underline;
}

.markdown-body .critic_comment {
  color: #8e908c;
  font-style: italic;
}

.markdown-body .headeranchor {
  font: normal normal 16px fontawesome-mini;
  line-height: 1;
  display: inline-block;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.headeranchor:before {
  content: '\e157';
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 4px 0.25em -20px;
  vertical-align: middle;
}

/* Media */
@media only screen and (min-width: 480px) {
  .markdown-body {
    font-size:14px;
  }
}

@media only screen and (min-width: 768px) {
  .markdown-body {
    font-size:16px;
  }
}

@media print {
  .markdown-body * {
    background: transparent !important;
    color: black !important;
    filter:none !important;
    -ms-filter: none !important;
  }

  .markdown-body {
    font-size:12pt;
    max-width:100%;
    outline:none;
    border: 0;
  }

  .markdown-body a,
  .markdown-body a:visited {
    text-decoration: underline;
  }

  .markdown-body .headeranchor-link {
    display: none;
  }

  .markdown-body a[href]:after {
    content: " (" attr(href) ")";
  }

  .markdown-body abbr[title]:after {
    content: " (" attr(title) ")";
  }

  .markdown-body .ir a:after,
  .markdown-body a[href^="javascript:"]:after,
  .markdown-body a[href^="#"]:after {
    content: "";
  }

  .markdown-body pre {
    white-space: pre;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  .markdown-body pre,
  .markdown-body blockquote {
    border: 1px solid #999;
    padding-right: 1em;
    page-break-inside: avoid;
  }

  .markdown-body .progress,
  .markdown-body .progress-bar {
    -moz-box-shadow: none;
    -webkit-box-shadow: none;
    box-shadow: none;
  }

  .markdown-body .progress {
    border: 1px solid #ddd;
  }

  .markdown-body .progress-bar {
    height: 22px;
    border-right: 1px solid #ddd;
  }

  .markdown-body tr,
  .markdown-body img {
    page-break-inside: avoid;
  }

  .markdown-body img {
    max-width: 100% !important;
  }

  .markdown-body p,
  .markdown-body h2,
  .markdown-body h3 {
    orphans: 3;
    widows: 3;
  }

  .markdown-body h2,
  .markdown-body h3 {
    page-break-after: avoid;
  }
}
</style><script type='text/javascript' src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js'></script><script>MathJax.Hub.Config({
  config: ["MMLorHTML.js"],
  extensions: ["tex2jax.js"],
  jax: ["input/TeX", "output/HTML-CSS", "output/NativeMML"],
  tex2jax: {
    inlineMath: [ ['$','$'], ["\\(","\\)"] ],
    displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
    processEscapes: true,
    processEnvironments: true,
    ignoreClass: ".*|",
    processClass: "arithmatex"
  },
  TeX: {
    extensions: ["AMSmath.js", "AMSsymbols.js"],
    TagSide: "right",
    TagIndent: ".8em",
    MultLineWidth: "85%",
    equationNumbers: {
      autoNumber: "AMS",
    },
    unicode: {
      fonts: "STIXGeneral,'Arial Unicode MS'"
    }
  },
  showProcessingMessages: false,
  messageStyle: 'none'
});
</script><style>/*GitHub*/
.codehilite {background-color:#fff;color:#333333;}
.codehilite .hll {background-color:#ffffcc;}
.codehilite .c{color:#999988;font-style:italic}
.codehilite .err{color:#a61717;background-color:#e3d2d2}
.codehilite .k{font-weight:bold}
.codehilite .o{font-weight:bold}
.codehilite .cm{color:#999988;font-style:italic}
.codehilite .cp{color:#999999;font-weight:bold}
.codehilite .c1{color:#999988;font-style:italic}
.codehilite .cs{color:#999999;font-weight:bold;font-style:italic}
.codehilite .gd{color:#000000;background-color:#ffdddd}
.codehilite .ge{font-style:italic}
.codehilite .gr{color:#aa0000}
.codehilite .gh{color:#999999}
.codehilite .gi{color:#000000;background-color:#ddffdd}
.codehilite .go{color:#888888}
.codehilite .gp{color:#555555}
.codehilite .gs{font-weight:bold}
.codehilite .gu{color:#800080;font-weight:bold}
.codehilite .gt{color:#aa0000}
.codehilite .kc{font-weight:bold}
.codehilite .kd{font-weight:bold}
.codehilite .kn{font-weight:bold}
.codehilite .kp{font-weight:bold}
.codehilite .kr{font-weight:bold}
.codehilite .kt{color:#445588;font-weight:bold}
.codehilite .m{color:#009999}
.codehilite .s{color:#dd1144}
.codehilite .n{color:#333333}
.codehilite .na{color:teal}
.codehilite .nb{color:#0086b3}
.codehilite .nc{color:#445588;font-weight:bold}
.codehilite .no{color:teal}
.codehilite .ni{color:purple}
.codehilite .ne{color:#990000;font-weight:bold}
.codehilite .nf{color:#990000;font-weight:bold}
.codehilite .nn{color:#555555}
.codehilite .nt{color:navy}
.codehilite .nv{color:teal}
.codehilite .ow{font-weight:bold}
.codehilite .w{color:#bbbbbb}
.codehilite .mf{color:#009999}
.codehilite .mh{color:#009999}
.codehilite .mi{color:#009999}
.codehilite .mo{color:#009999}
.codehilite .sb{color:#dd1144}
.codehilite .sc{color:#dd1144}
.codehilite .sd{color:#dd1144}
.codehilite .s2{color:#dd1144}
.codehilite .se{color:#dd1144}
.codehilite .sh{color:#dd1144}
.codehilite .si{color:#dd1144}
.codehilite .sx{color:#dd1144}
.codehilite .sr{color:#009926}
.codehilite .s1{color:#dd1144}
.codehilite .ss{color:#990073}
.codehilite .bp{color:#999999}
.codehilite .vc{color:teal}
.codehilite .vg{color:teal}
.codehilite .vi{color:teal}
.codehilite .il{color:#009999}
.codehilite .gc{color:#999;background-color:#EAF2F5}
</style><title>README</title></head><body><article class="markdown-body"><h1 id="xenoturbella-single-cell-rna-seq-analysis"><em>Xenoturbella</em> single-cell RNA-seq analysis<a class="headerlink" href="#xenoturbella-single-cell-rna-seq-analysis" title="Permanent link"></a></h1>
<p>The <em>Xenoturbella</em> single-cell RNA analysis pipeline is set up as a
series of scripts, forming an experiment.</p>
<h3 id="installation">Installation<a class="headerlink" href="#installation" title="Permanent link"></a></h3>
<p>The project makes use of R and Python 3 (for scScope, RNA velocity). To
install required packages, run the scripts with active interpreters or
virtual environments.</p>
<p>R:</p>
<div class="codehilite"><pre>Rscript requirements.R
</pre></div>


<p>Python:</p>
<div class="codehilite"><pre>pip install -r requirements.txt
</pre></div>


<h3 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link"></a></h3>
<p>An experiment is identified
by a name and is stored in the output folder under a single directory.
Optionally, the maximum number of cells per library can be specified
- experiments on reduced data can be used for debugging.
The data subset used in this study can comprise all libraries
(<code>DATA=all</code>) or the two whole animals (<code>DATA=mix</code>).</p>
<p>Example values for debug:</p>
<div class="codehilite"><pre>EXP=test_experiment_all
MAX_CELLS=1000
DATA=all
</pre></div>


<p>The code is meant to be run in the project root directory, but can
be easily modified to do otherwise. Global paths to data
and code are set in <code>./scripts/utils.R</code> .</p>
<div class="codehilite"><pre># Set base paths
out.base.dir = &quot;output&quot;
in.base.dir = &quot;data&quot;
in.script.dir = &quot;scripts&quot;

# Outputs
in.dir = file.path(in.base.dir, &quot;matrix&quot;)
in.h5ad.dir = file.path(in.base.dir, &quot;h5ad&quot;)

# Marker sheets
in.models = file.path(in.base.dir, &quot;markers&quot;, &quot;Xenotubella_gene_models_all.csv&quot;)
in.markers = file.path(in.base.dir, &quot;markers&quot;, &quot;Xenoturbella_protein_gene_markers.csv&quot;)
in.neuro.markers = file.path(in.base.dir, &quot;markers&quot;, &quot;Xenoturbella_neurotransmitters.csv&quot;)
</pre></div>


<p>Some analyses require to uncompress the following large files inside the <code>data/</code> directory.</p>
<div class="codehilite"><pre>- `matrix.tar.gz` (Main analysis)
- `h5ad.tar.gz ` (scScope imputation and batch effect correction)
- `loom.tar.gz` (RNA velocity)
</pre></div>


<h3 id="loading-own-files">Loading own files<a class="headerlink" href="#loading-own-files" title="Permanent link"></a></h3>
<p>To run the analysis pipeline on custom data sets, modify the
<code>xeno.load.design</code> function, which constructs should return
a data frame with columns</p>
<ul>
<li><code>path</code>: Name of directory in 10x format in <code>matrix/</code>.</li>
<li><code>name</code>: Human-readable name of the library.</li>
<li><code>color</code>: Color of library to be used in plots.</li>
<li><code>animal</code>: Optional &ldquo;batch&rdquo; parameter.</li>
</ul>
<p>Default definition:</p>
<div class="codehilite"><pre># Define experimental design
xeno.load.design &lt;- function(dset=&quot;all&quot;){
  design = data.frame(path=c(
    &quot;Xenoturbella_dorsal_matrix&quot;,
    &quot;Xenoturbella_mix1_matrix&quot;,
    &quot;Xenoturbella_mix2_matrix&quot;,
    &quot;Xenoturbella_skin_matrix&quot;,
    &quot;Xenoturbella_ventral_1_5hr_matrix&quot;,
    &quot;Xenoturbella_ventral_2hr_matrix&quot;),
      animal = c(&quot;animalC&quot;,
                 &quot;animalAB&quot;,
                 &quot;animalAB&quot;,
                 &quot;animalC&quot;,
                 &quot;animalC&quot;,
                 &quot;animalC&quot;),
      color = c(&quot;#f8766d&quot;, &quot;#b79f00&quot;, &quot;#00ba38&quot;, &quot;#00bfc4&quot;, &quot;#619cff&quot;, &quot;#f564e3&quot;),
    stringsAsFactors = FALSE)
  design$name = unlist(lapply(strsplit(design$path, &quot;_&quot;),
                              FUN = function(x){paste(x[2:3], collapse = &quot;_&quot;)}))

  if(dset != &quot;all&quot;){
    design = design[grep(dset, design$path),]
  }
  row.names(design) = design$name
  return(design)
}
</pre></div>


<h2 id="pre-processing-qc-and-basic-visualization">Pre-processing, QC and basic visualization<a class="headerlink" href="#pre-processing-qc-and-basic-visualization" title="Permanent link"></a></h2>
<p>Scripts related to basic pre-processing and data display.</p>
<h3 id="scscope-data-imputation-and-batch-effect-removal">scScope data imputation and batch effect removal<a class="headerlink" href="#scscope-data-imputation-and-batch-effect-removal" title="Permanent link"></a></h3>
<p>The data sparsity and batch effect provide the major challenges in this
study.  The <a href="https://github.com/AltschulerWu-Lab/scScope">scScope</a>
autoencoder model is used to correct the gene expression matrix for
these confounding factors. The autoencoder depth and the number of latent
 factors was chosen according to the methods section (see paper).
 The resulting files with imputed gene expression matrix
 are readily available in this repository (<code>./data/h5ad/scScope__all.h5a</code>).</p>
<div class="codehilite"><pre>IN=&quot;./data/h5ad/all.h5ad&quot;
OUT_DIR=&quot;./data/h5ad/all.h5ad&quot;
DEPTH=128
LATENT=50
./scScope.py --input $IN --output-dir $OUT_DIR --batch-var orig.ident
--n-latent $LATENT --depth-encoder $DEPTH --depth-decoder $DEPTH
</pre></div>


<h3 id="basic-quality-control">Basic quality control<a class="headerlink" href="#basic-quality-control" title="Permanent link"></a></h3>
<p>This scripts counts the number of genes per cell, computes distribution
of detected genes per cell and gene expression matrix density. An example
run</p>
<div class="codehilite"><pre>./scripts/main/basic-qc.R -e $EXP -m $MAX_CELLS
</pre></div>


<p>Stores the output in <code>./output/test_experiment/basic_qc/</code>.</p>
<p>The parameter specification for each script is obtained with <code>-h</code> switch.</p>
<div class="codehilite"><pre>./scripts/main/basic-qc.R -h
</pre></div>


<p>prints:</p>
<div class="codehilite"><pre>Usage: ./scripts/main/basic-qc.R [options]
Basic quality control script. Count the number of genes/cell and gene expression matrix density.

Options:
    -e EXPERIMENT, --experiment=EXPERIMENT
        Experiment name.

    -m MAX, --max=MAX
        Maximum number of cells to load per library.

    -c CONFIG, --config=CONFIG
        Configuration script path (scripts/utils.R)

    -h, --help
        Show this help message and exit
</pre></div>


<h3 id="dimensionality-reduction-and-visualization">Dimensionality reduction and visualization<a class="headerlink" href="#dimensionality-reduction-and-visualization" title="Permanent link"></a></h3>
<p>This basic script constructs the central Seurat data object and
performs one or more dimensionality reductions: <code>pca</code>, <code>cca</code>
or <code>scscope</code>. Each dimensionality reduction method is the used
by t-SNE to produce a 2D visualization.</p>
<div class="codehilite"><pre>./scripts/main/preprocess.R -e $EXP -m $MAX_CELLS -r scscope
</pre></div>


<p>Produces</p>
<div class="codehilite"><pre>output/test_experiment/pca/tsne_plot.pdf
output/test_experiment/scscope/tsne_plot.pdf
</pre></div>


<p>The R data object to be used by downstream scripts is stored in
<code>output/test_experiment/data.Robj</code>.</p>
<h3 id="gene-expression-overlay">Gene expression overlay<a class="headerlink" href="#gene-expression-overlay" title="Permanent link"></a></h3>
<p>Augment 2D visualization with gene expression plots. Gene markers
are read from <code>data/markers/Xenoturbella_protein_gene_markers.csv</code></p>
<div class="codehilite"><pre> ./scripts/main/features.R -e $EXP
</pre></div>


<p>Produces marker gene expression plots for all dimensionality reductions
performed previously.</p>
<div class="codehilite"><pre>output/test_experiment/pca/markers/Cell_Adhesion.tiff
output/test_experiment/pca/markers/Innexins.tiff
output/test_experiment/pca/markers/Ion_Channels.tiff
output/test_experiment/pca/markers/Neuropeptides.tiff
output/test_experiment/pca/markers/Neurotransmitter_Biosynthesis.tiff
</pre></div>


<p>Example visualization of Muscle cells:</p>
<p><img alt="Muscle" src="/Users/martins/Dev/xenoturbella/output/test_experiment/pca/markers/Muscle.1.tiff" /></p>
<h2 id="clustering-and-cluster-analysis">Clustering and cluster analysis<a class="headerlink" href="#clustering-and-cluster-analysis" title="Permanent link"></a></h2>
<p>In the following, we devise methods for finding clusters of cells in <em>Xenoturbella</em>.</p>
<h3 id="identifying-clusters-of-cells-by-louvain-clustering">Identifying clusters of cells by Louvain clustering<a class="headerlink" href="#identifying-clusters-of-cells-by-louvain-clustering" title="Permanent link"></a></h3>
<p>The potential number of different cell clusters - cell types - are identified by Louvain clustering.  This method automatically find the number of clusters using modularity optimization. It however depends on a resolution hyperparameter, governing the minimal distance between for two cells to be deemed similar. Here, we try <em>N</em> uniformly separated values of the resolution between empirically selected 0.1 and 10.1. </p>
<div class="codehilite"><pre>REDC=pca
N=10
./scripts/clustering/clustering-louvain.R -e $EXP -n $N -r $REDC
</pre></div>


<p>For each configuration, the script computes the clustering outcome: clustering assignment and a t-SNE visualization augmented with cluster identity.</p>
<div class="codehilite"><pre>output/test_experiment/clustering/pca/1/tsne_plot_orig.ident_leg-TRUE.pdf
output/test_experiment/clustering/pca/1/tsne_plot_ident_leg-TRUE.pdf
...
output/test_experiment/clustering/pca/20/tsne_plot_orig.ident_leg-TRUE.pdf
output/test_experiment/clustering/pca/20/tsne_plot_ident_leg-TRUE.pdf
</pre></div>


<p>Clustering assignments for different configurations are stored in:</p>
<div class="codehilite"><pre>output/test_experiment/clustering/pca/
output/test_experiment/clustering/pca/
</pre></div>


<h3 id="choosing-the-number-of-clusters-with-a-second-experiment">Choosing the number of clusters with a second experiment<a class="headerlink" href="#choosing-the-number-of-clusters-with-a-second-experiment" title="Permanent link"></a></h3>
<p>We use a biologically-inspired strategy to decide on the number of clusters. Namely, we perform another experiment, consisting of only cells from the two whole animals (libraries labelled <em>mix</em>). Intuitively, the two whole animal samples should contain similar cell types as the four individual parts, and the cell types should be present in roughly consistent proportions. This strategy disregards and experimental/technical biases.</p>
<p>Suppose we run another preprocessing and clustering experiment, this time selecting only the cells from the two whole animals.</p>
<div class="codehilite"><pre>EXP2=test_experiment_mix
DATA2=mix
MAX_CELLS=2000
REDC=pca
./scripts/main/preprocess.R -e $EXP2 -d $DATA2 -m $MAX_CELLS -r $REDC
./scripts/clustering/clustering-louvain.R -e $EXP2 -n $N -r $REDC
</pre></div>


<p>This second experiment (<code>test_experiment_mix</code>) is then used to select the number of clusters in our main experiment (<code>test_experiment_all</code>). Briefly, the two experiments are combined to find a clustering configuration that is consistent in both experiments - having a similar number of clusters and similar cluster size distribution (see paper). The chosen clustering assignment is stored in the R object corresponding to the first experiment (<code>-e $EXP</code>).</p>
<div class="codehilite"><pre>./scripts/clustering/clustering-analysis.R -e $EXP -E $EXP2 -r $REDC
</pre></div>


<p>In this artificial case, we decide to select X clusters, supported by both experiments.</p>
<h3 id="tree-of-clusters">Tree of clusters<a class="headerlink" href="#tree-of-clusters" title="Permanent link"></a></h3>
<p>The clusters can be represented using hierarchical clustering with significance (obtained by package <a href="">pvclust</a>) to distinguish significant versus non-significant clusters of cells.</p>
<p>The following dimensionality reduction are available.</p>
<ul>
<li><strong>hvg</strong>: Highly variable genes.</li>
<li><strong>scscope</strong>, <strong>pca</strong>, <strong>cca</strong>: Output of scScope, PCA or CCA, respectively.</li>
</ul>
<p>The script is run as follows.</p>
<div class="codehilite"><pre>./scripts/clustering/clustering-pvclust.R -e $EXP -r $REDC
</pre></div>


<p>The clustering tree is stored as follows.</p>
<div class="codehilite"><pre>output/test_experiment_all/clustering/pca/clustering_4_pvclust.pdf
output/test_experiment_all/clustering/pca/tree.Robj
</pre></div>


<h3 id="tree-and-clusters-of-neurons">Tree and clusters of neurons<a class="headerlink" href="#tree-and-clusters-of-neurons" title="Permanent link"></a></h3>
<p>A similar visualization is used to show relations between potential neurons in the dataset, stratified by the expressing neurotransmitter molecule. The following script also computes a matching t-SNE visualization that emphasizes the clusters and the library of origin.</p>
<p>Dimensionality reductions are the same as in the tree of clusters (above).</p>
<div class="codehilite"><pre>./scripts/clustering/clustering-pvclust-neurons.R -e $EXP -r $REDC
</pre></div>


<p>The output files display three t-SNE plots and one hierarchical clustering tree.</p>
<div class="codehilite"><pre>output/test_experiment_all/clustering/pca-neurons/clustering_5_pvclust.pdf
output/test_experiment_all/clustering/pca-neurons/combined-tsne-cluster.pdf
output/test_experiment_all/clustering/pca-neurons/combined-tsne-neurotransmitter.pdf
output/test_experiment_all/clustering/pca-neurons/combined-tsne-sample.pdf
</pre></div>


<h2 id="differential-gene-expression">Differential gene expression<a class="headerlink" href="#differential-gene-expression" title="Permanent link"></a></h2>
<p>Differential gene expression between clusters, super-clusters (see paper) and neurons.</p>
<h3 id="differential-expression-test">Differential expression test<a class="headerlink" href="#differential-expression-test" title="Permanent link"></a></h3>
<p>The test type (<code>--type $TYPE</code>) determines that data subset and stratification to be tested:</p>
<ul>
<li><strong>clusters</strong>: Differentially expressed genes between cell clusters.</li>
<li><strong>neurons</strong>: Test between 6 neurotramsmitters and all other cells.</li>
<li><strong>neurons-specific</strong>: Test specifically for the cells expressing the neurotransmitters.</li>
<li><strong>tree</strong>: Test between different branches of clustering tree. The parameter <code>--tree-start</code> must be set to determine at which cluster to split the tree.</li>
<li><strong>superclusters</strong>: Test between empirically determined cell types: extracellular, primed and undifferentiated.</li>
<li><strong>parts</strong>: Test between six libraries.</li>
</ul>
<p>The test is run as follows:</p>
<div class="codehilite"><pre>TYPE=&quot;clusters&quot;
MAX_GENES=100
./scripts/de/diffexp.R -e $EXP -M $MAX_GENES -t $TYPE
</pre></div>


<p>The result is a differential gene expression table <code>output/test_experiment_all/diffexp-clusters/diffexp.csv</code>, with one gene per row.</p>
<ul>
<li><strong>geneID</strong>: Gene identifier.  </li>
<li><strong>n_cluster</strong>: Number of cells in the group (cluster).   </li>
<li><strong>n_other</strong>: Number of other cells. </li>
<li><strong>nexp_cluster</strong>: Number cells in the cluster expressing the gene. </li>
<li><strong>nexp_other</strong>: Number of other cells expressing the gene.   </li>
<li><strong>pexp_cluster</strong>: Fraction cells in the cluster expressing the gene. </li>
<li><strong>pexp_other</strong>: Fraction of other cells expressing the gene.   </li>
<li><strong>lfc</strong>: Log-fold change between cluster and other cells. </li>
<li><strong>pvalue</strong>: Significance of the test result.  </li>
<li><strong>fdr</strong>: False discovery rate, Benjamini-Hochberg corrected p-value. </li>
<li><strong>test.type</strong>: Selected test type.   </li>
<li><strong>cluster</strong>: Cluster identifier. </li>
<li><strong>marker</strong>: Is the gene a previously known marker (File <code>in.markers</code>).  </li>
<li><strong>annotation</strong>: Gene annotation.</li>
</ul>
<h3 id="differential-expression-heatmaps">Differential expression heatmaps<a class="headerlink" href="#differential-expression-heatmaps" title="Permanent link"></a></h3>
<p>The results of a given differential expression can be shown in a heatmap. 
Dimesionality reduction is used to load a clustering tree, if available. </p>
<div class="codehilite"><pre>TYPE=&quot;clusters&quot;
N_GENES=1
./scripts/de/heatmap.R -e $EXP -M $N_GENES -t $TYPE -r $REDC
</pre></div>


<p>The output files are gene expression heatmaps in <code>output/test_experiment_all/diffexp-clusters/heatmaps/</code>.</p>
<p><img alt="heatmap" src="/Users/martins/Dev/xenoturbella/output/test_experiment_all/diffexp-clusters/heatmaps/cluster_groups_top_expression_1.tiff" /></p>
<h2 id="rna-velocity">RNA velocity<a class="headerlink" href="#rna-velocity" title="Permanent link"></a></h2>
<p>The RNA velocity is computed on a given dimensionality reduction. The velocity estimates
are obtained with the <code>scvelo</code> package. A <code>.loom</code> file containing all six libraries is 
prepared with this repository.</p>
<div class="codehilite"><pre>LOOM=data/loom/xeno-combined-bc.loom
</pre></div>


<p>The script computes RNA velocity estimates and stores them in a pickled object, along with 
the transition matrix. The velocity vectors are embedded on a chosen dimensionality reduction (t-SNE) plot.</p>
<div class="codehilite"><pre>./scripts/velocity/velocity_compute.py -e $EXP -L $LOOM -r $REDC
</pre></div>


<p>The resulting objects are:</p>
<div class="codehilite"><pre>output/test_experiment_all/velocity/transitions.npz
output/test_experiment_all/velocity/data.pkl.gz
</pre></div>


<p>To overlay the velocity plot with the marker genes (stored in <code>metadata.csv</code>), run</p>
<div class="codehilite"><pre>./scripts/velocity/velocity_markers.py -e $EXP
</pre></div>


<p>The result are RNA velocity plots embedded into a chosed t-SNE and overlayed with marker genes expression.</p>
<div class="codehilite"><pre>output/test_experiment_all/velocity/markers/velocity_nGene.tiff
output/test_experiment_all/velocity/markers/velocity_nGene_no_legend.tiff
output/test_experiment_all/velocity/markers/velocity_nUMI.tiff
output/test_experiment_all/velocity/markers/velocity_nUMI_no_legend.tiff
output/test_experiment_all/velocity/markers/velocity_orig.ident.tiff
output/test_experiment_all/velocity/markers/velocity_orig.ident_no_legend.tiff
</pre></div>


<p>Example output (Library of origin, <code>orig.ident</code>):
<img alt="velocity-example" src="/Users/martins/Dev/xenoturbella/output/test_experiment_all/velocity/markers/velocity_orig.ident.tiff" /></p>
<h3 id="r-session-info">R session info<a class="headerlink" href="#r-session-info" title="Permanent link"></a></h3>
<div class="codehilite"><pre>&gt; sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
 [1] RColorBrewer_1.1-2          Rtsne_0.15                  ramify_0.3.3                pracma_2.2.2
 [5] pheatmap_1.0.12             MAST_1.8.2                  SingleCellExperiment_1.4.1  SummarizedExperiment_1.12.0
 [9] DelayedArray_0.8.0          BiocParallel_1.16.0         matrixStats_0.54.0          Biobase_2.42.0
[13] GenomicRanges_1.34.0        GenomeInfoDb_1.18.1         IRanges_2.16.0              S4Vectors_0.20.1
[17] BiocGenerics_0.28.0         cluster_2.0.7-1             Seurat_2.3.4                Matrix_1.2-15
[21] cowplot_0.9.3               ggplot2_3.1.0               pvclust_2.0-0

loaded via a namespace (and not attached):
  [1] colorspace_1.3-2       class_7.3-14           modeltools_0.2-22      ggridges_0.5.1         mclust_5.4.2
  [6] htmlTable_1.12         XVector_0.22.0         base64enc_0.1-3        rstudioapi_0.8         proxy_0.4-22
 [11] npsurv_0.4-0           flexmix_2.3-14         bit64_0.9-7            mvtnorm_1.0-8          codetools_0.2-15
 [16] splines_3.5.1          R.methodsS3_1.7.1      lsei_1.2-0             robustbase_0.93-3      knitr_1.20
 [21] Formula_1.2-3          jsonlite_1.5           ica_1.0-2              kernlab_0.9-27         png_0.1-7
 [26] R.oo_1.22.0            compiler_3.5.1         httr_1.3.1             backports_1.1.2        assertthat_0.2.0
 [31] lazyeval_0.2.1         lars_1.2               acepack_1.4.1          htmltools_0.3.6        tools_3.5.1
 [36] igraph_1.2.2           GenomeInfoDbData_1.2.0 gtable_0.2.0           glue_1.3.0             RANN_2.6
 [41] reshape2_1.4.3         dplyr_0.8.0.1          Rcpp_1.0.1             trimcluster_0.1-2.1    gdata_2.18.0
 [46] ape_5.2                nlme_3.1-137           iterators_1.0.10       fpc_2.1-11.1           gbRd_0.4-11
 [51] lmtest_0.9-36          stringr_1.3.1          irlba_2.3.3            gtools_3.8.1           DEoptimR_1.0-8
 [56] zlibbioc_1.28.0        MASS_7.3-51.1          zoo_1.8-4              scales_1.0.0           doSNOW_1.0.16
 [61] yaml_2.2.0             reticulate_1.10        pbapply_1.3-4          gridExtra_2.3          rpart_4.1-13
 [66] segmented_0.5-3.0      latticeExtra_0.6-28    stringi_1.2.4          foreach_1.4.4          checkmate_1.8.5
 [71] caTools_1.17.1.1       bibtex_0.4.2           Rdpack_0.10-1          SDMTools_1.1-221       rlang_0.3.4.9003
 [76] pkgconfig_2.0.2        dtw_1.20-1             prabclus_2.2-6         bitops_1.0-6           lattice_0.20-38
 [81] ROCR_1.0-7             purrr_0.2.5            htmlwidgets_1.3        bit_1.1-14             tidyselect_0.2.5
 [86] plyr_1.8.4             magrittr_1.5           R6_2.3.0               snow_0.4-3             gplots_3.0.1
 [91] Hmisc_4.1-1            pillar_1.3.1           foreign_0.8-71         withr_2.1.2            fitdistrplus_1.0-11
 [96] mixtools_1.1.0         abind_1.4-5            RCurl_1.95-4.11        survival_2.43-1        nnet_7.3-12
[101] tibble_2.1.1           tsne_0.1-3             crayon_1.3.4           hdf5r_1.0.1            KernSmooth_2.23-15
[106] grid_3.5.1             data.table_1.11.8      metap_1.0              digest_0.6.18          diptest_0.75-7
[111] tidyr_0.8.2            R.utils_2.7.0          munsell_0.5.0
</pre></div></article></body></html>